tests/ci-helper.test.ts

```diff
@@ -1,11 +1,132 @@
+    return { smtpUser, smtpHost, smtpPass, smtpOpts };
+}
+
+// Mocking class to replace GithubGlue with mock of GitHubGlue
+
+class TestCIHelper extends CIHelper {
+public ghGlue: GitHubGlue;          // not readonly reference
+    public addPRComment: any;
+
+    public constructor(workDir?: string, debug = false, gggDir = ".") {
+        super(workDir, debug, gggDir);
+        this.testing = true;
+        this.ghGlue = this.github;
+        this.addPRComment = jest.fn();
+        this.ghGlue.addPRComment = this.addPRComment;
+        // need keys to authenticate
+        // this.ghGlue.ensureAuthenticated = async (): Promise<void> => {};
+    }
+
+    public setGHgetPRInfo(o: IPullRequestInfo) {
+        this.ghGlue.getPRInfo = jest.fn( async ():
+            Promise<IPullRequestInfo> => o );
+    }
+
+    public setGHgetPRComment(o: IPRComment) {
+        this.ghGlue.getPRComment = jest.fn( async ():
+            Promise<IPRComment> => o );
+    }
+
+    public setGHgetGitHubUserInfo(o: IGitHubUser) {
+        this.ghGlue.getGitHubUserInfo = jest.fn( async ():
+            Promise<IGitHubUser> => o );
+    }
+}
+
+// Create three repos.
+// worktree is a local copy for doing updates and has the config
+// info that would normally be in the gitgitgadget repo.  To ensure
+// testing isolation, worktree is NOT the repo used for git clone
+// tests.  That work is done in gggLocal.
+
+// gggRemote represents the master on github.
+
+// gggLocal represents the empty repo to be used by gitgitgadget.  It
+// is empty to ensure nothing needs to be present (worktree would
+// have objects present).
+
+async function setupRepos(instance: string):
+    Promise <{ worktree: TestRepo, gggLocal: TestRepo, gggRemote: TestRepo }> {
+    const worktree = await testCreateRepo(__filename, `-work-cmt${instance}`);
+    const gggLocal = await testCreateRepo(__filename, `-git-lcl${instance}`);
+    const gggRemote = await testCreateRepo(__filename, `-git-rmt${instance}`);
+
+    // re-route the URLs
+    await worktree.git(["config", `url.${gggRemote.workDir}.insteadOf`,
+                        "https://github.com/gitgitgadget/git"]);
+
+    await gggLocal.git(["config", `url.${gggRemote.workDir}.insteadOf`,
+                        "https://github.com/gitgitgadget/git"]);
+
+    // set needed config
+    await worktree.git(["config",
+        '--add', "gitgitgadget.workDir", gggLocal.workDir]);
+    await worktree.git(["config",
+        '--add', "gitgitgadget.publishRemote",
+        "https://github.com/gitgitgadget/git"]);
+
+    const { smtpUser, smtpHost, smtpPass, smtpOpts } =
+        await getSMTPInfo();
+
+    await worktree.git(["config",
+        '--add', "gitgitgadget.smtpUser", smtpUser ? smtpUser : "test"]);
+
+    await worktree.git(["config",
+        '--add', "gitgitgadget.smtpHost", smtpHost ? smtpHost : "test"]);
+
+    await worktree.git(["config",
+        '--add', "gitgitgadget.smtpPass", smtpPass ? smtpPass : "test"]);
```

Comment:
You could have mentioned this when I was whining about config issues. 😀 I copied from gitgitgadget.test.ts.  Can we defer changing to use `GIT_CONFIG_PARAMETERS` for now?  That will take some extra testing to ensure there are no cross-platform issues with setting the env var (any possible escaping that may be needed for example).

btw, the CI test setup will need to be updated to have the last two tests actually run.  They will add quite a bit to the test run-time.  I created an account with ethereal and it works well.  They are deleting emails faster than originally documented but it is good for testing.

Reply:
> You could have mentioned this when I was whining about config issues. 😀

Sorry, I did not think of it earlier...

> I copied from gitgitgadget.test.ts. Can we defer changing to use `GIT_CONFIG_PARAMETERS` for now?

Absolutely. We do not even have to do it. The current way (using the repository config) is definitely more robust.

> That will take some extra testing to ensure there are no cross-platform issues with setting the env var (any possible escaping that may be needed for example).

Right. It might not even be a good idea to use `GIT_CONFIG_PARAMETERS`; It would be more of a hack than a solid solution, I think.

> btw, the CI test setup will need to be updated to have the last two tests actually run. They will add quite a bit to the test run-time. I created an account with ethereal and it works well. They are deleting emails faster than originally documented but it is good for testing.

You mean the part guarded by `if (smtpUser) {                 // if configured for this test`? That's a good point. I registered an account, and added the `gitgitgadget.CIsmtp*` variables to the pipeline, but as one of them is secret, it won't show up in the environment variables, therefore I cannot use above-mentioned `GIT_CONFIG_PARAMETERS` trick to "magically" inject them into the test run.

I think what we need is something like this in the `azure-pipelines.yml` file (before the `npm` step):

```yml
- script: |
    git config gitgitgadget.CIsmtpUser $(gitgitgadget.CIsmtpUser)
    git config gitgitgadget.CIsmtpHost $(gitgitgadget.CIsmtpHost)
    git config gitgitgadget.CIsmtpPass $(gitgitgadget.CIsmtpPass)
    git config gitgitgadget.CIsmtpOpts $(gitgitgadget.CIsmtpOpts)
  displayName: 'Configure ethereal SMTP'
  condition: ne(variables['System.PullRequest.IsFork'], 'True')
```

I have not tested this, of course, and it might be better (if it works) to use the condition

```
  condition: ne(variables['gitgitgadget.CIsmtpPass'], '')
```

Reply:
> Sorry, I did not think of it earlier...

Well you did mention it somewhere along the way and I missed it (probably information overload).  It is mentioned now in the wiki page on pipelines.

> I think what we need is something like this in the `azure-pipelines.yml` file (before the `npm` step):
> 
> ```yaml
> - script: |
>     git config gitgitgadget.CIsmtpUser $(gitgitgadget.CIsmtpUser)
>     git config gitgitgadget.CIsmtpHost $(gitgitgadget.CIsmtpHost)
>     git config gitgitgadget.CIsmtpPass $(gitgitgadget.CIsmtpPass)
>     git config gitgitgadget.CIsmtpOpts $(gitgitgadget.CIsmtpOpts)
>   displayName: 'Configure ethereal SMTP'
>   condition: ne(variables['System.PullRequest.IsFork'], 'True')
> ```
> 
> I have not tested this, of course, and it might be better (if it works) to use the condition
> 
> ```
>   condition: ne(variables['gitgitgadget.CIsmtpPass'], '')
> ```

I can work on this if you want.  Obviously I can't set the pipeline vars but i could get it to the point of Azure not complaining about the YAML.

Reply:
> Obviously I can't set the pipeline vars

I already configured those variables yesterday ;-)

> but I could get it to the point of Azure not complaining about the YAML.

Let me give it a shot

Reply:
Worked: See the [manually triggered build](https://dev.azure.com/gitgitgadget/git/_build/results?buildId=25364&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=5caf77c8-9b10-50ef-b5c7-ca89c63e1c86) (which did run the SMTP tests, and I was able to verify that by logging into https://ethereal.email/, too) and the [PR build](https://dev.azure.com/gitgitgadget/git/_build/results?buildId=25363&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9) (which skipped the SMTP configuration and hence the SMTP tests).